lxd/storage/drivers/driver_zfs_volumes.go

```diff
@@ -2851,14 +2871,25 @@ func (d *zfs) MountVolumeSnapshot(snapVol Volume, op *operations.Operation) erro
 			fsVol := snapVol.NewVMBlockFilesystemVolume()
 			err = d.MountVolumeSnapshot(fsVol, op)
 			if err != nil {
-				return err
+				return nil, err
 			}
 		}
 	}
 
-	snapVol.MountRefCountIncrement() // From here on it is up to caller to call UnmountVolumeSnapshot() when done.
+	d.logger.Debug("Mounted ZFS snapshot dataset", logger.Ctx{"dev": snapshotDataset, "path": mountPath})
+
+	revert.Add(func() {
+		_, err := forceUnmount(mountPath)
+		if err != nil {
+			return
+		}
+
+		d.logger.Debug("Unmounted ZFS snapshot dataset", logger.Ctx{"dev": snapshotDataset, "path": mountPath})
+	})
+
+	cleanup := revert.Clone().Fail
 	revert.Success()
-	return nil
+	return cleanup, nil
```

Comment:
Why the need to return the cleanup hook here rather than calling `UnmountVolumeSnapshot()` when done?

Reply:
When calling `readonlySnapshot` the cleanup should be done by the caller. That's the way it was, and I didn't change that.